src/project-roots.ts

```diff
@@ -7,7 +7,18 @@ const walkSync = require('walk-sync');
 
 export class Project {
 
-  constructor(public readonly root: string) {
+  public readonly root: string;
+  public readonly podRoot: string = '';
+
+  constructor(root: string) {
+
+    let env = require(join(root, 'config', 'environment.js'))();
```

Comment:
Yeah, this was the part I was most uncertain and concerned about. Do you have any suggestions? I'm not very familiar with fs caching etc. 

The only thing coming to mind is to circumvent the issue entirely by having this be something explicitly configured (probably in the vscode-ember extension?).

Reply:
as long as this value is not statically analyzeable and requires a JS file to be evaluated there is not much we can do...

I'd suggest to support pods without the prefix for now and once we find a proper solution we can try to suggest the prefix cases too

Reply:
:/ I understand, though that would be frustrating, because it would make this upgrade totally useless to my app.  Also, supporting pods without prefixes means showing invalid completions for prefixed ones. E.g., `pods/route-name/route.js` will still be detected and shown as `pods.route-name`

Possible alternative, though I'm not sure about efficiency: Files in pods have predefined filenames, e.g., `route-name/route.js`, `c-name/component.js`, `route-name/template.hbs`. If we use looser globbing, we don't even have to care about what the root pod directory is. 

So we could do something like:
```ts 
  safeWalkSync(join(root, 'app'), {
    globs: [
      '*/**/route.js',
      '*/**/components/**/component.js',
      '*/**/components/**/template.hbs'
    ]
  });
```
Then we'd have to strip the pod directory from the completion, which is easy for `*/components`, but more difficult for routes.

But it does mean working completions and definitions for components.

Reply:
I assume `pods` is the default prefix for pods? Maybe we could support no-prefix and default-prefix?

Reply:
I don't know if it's the default, but it certainly [seems to be the most common, by far](https://github.com/search?p=1&q=podModulePrefix+filename%3Aenvironment%2Cjs+path%3Aconfig&type=Code). But anyone using another prefix would still see invalid completions, as well as anyone using 'pods' as a route name ([and there are some](https://github.com/search?q=ember+router+pods+filename%3Arouter&type=Code)). It seems somewhat fragile.

Looking a little into it: it seems relatively trivial to watch `environment.js` (either via `workspace.createFileSystemWatcher` in the client extension or `fs.watchFile` on the server), delete the require cache and then re-require it. In my primitive testing, re-requiring the file took ~1ms, and the most complex environment.js files that I copied from github took ~2-3ms. Is that too slow? 

The other alternative is adding a config option in the `vscode-ember` client which could then be passed to the server.

I don't think supporting pods without a prefix or a default prefix is viable, simply because we won't be able to detect a non-supported configuration and anyone using those will get invalid completions.

Reply:
the other issue with that file (that I unfortunately forgot to mention before) is that it expects to be passed an environment, and depending on the environment and potentially other things it might change the prefixes.

and I have also seen projects where the file has significantly more logic in it and e.g. shells out to `git` to get the current commit hash etc.

@t-sauer any opinions?

Reply:
If it is a big problem reading the pod prefix from environment.js, I could live with having to specify it a second time.